### PR TITLE
fix: virtualize pool

### PR DESCRIPTION
**Pull request by erights**
Virtualize pool. Unlike #4707, this one does not virtualize vpool (the single pool) because it seems that the vpool need not be made persistent.

See https://github.com/Agoric/agoric-sdk/pull/4707#discussion_r831615803

Note to reviewers: I may still be confused. I suspect that @Chris-Hibbert and I need to pair program this before it is good.

### CODE REVIEW

**Comment by Chris-Hibbert** [comment]
I'm expecting to see a test demonstrating that objects can be defined, saved, flushed from memory, and restored with behavior identical to the original. Is there a reason that's not appropriate here?

**Comment by erights** [comment]
At https://github.com/Agoric/agoric-sdk/pull/4706#issuecomment-1055681522 @Chris-Hibbert asks:

> I'm expecting to see a test demonstrating that objects can be defined, saved, flushed from memory, and restored with behavior identical to the original. Is there a reason that's not appropriate here?

@FUDCo , should this PR test that? If so, how? Thanks.

**Comment by erights** [comment]
Converting back to draft because it is only a step towards #4707 . I'll let you know when #4707 is ready for review.

**Comment by erights** [comment]
Given @Chris-Hibbert 's comments at https://github.com/Agoric/agoric-sdk/pull/4707#discussion_r831615803 I put #4707  into draft status. Instead, it seems this PR #4706 is closer to what we want, so I'm declaring #4706  Ready to Review. Everyone, please review this one instead.

**Comment by erights** [comment]
Changed my mind again. Back to Draft until @Chris-Hibbert and I pair on this and it becomes ready. No need for others to review before then.

**Comment by warner** [comment]
> At [#4706 (comment)](https://github.com/Agoric/agoric-sdk/pull/4706#issuecomment-1055681522) @Chris-Hibbert asks:
> 
> > I'm expecting to see a test demonstrating that objects can be defined, saved, flushed from memory, and restored with behavior identical to the original. Is there a reason that's not appropriate here?
> 
> @FUDCo , should this PR test that? If so, how? Thanks.

We don't have a way to exercise that until #1848 / #4382 is done and we can upgrade vats. We've got test coverage on durable objects/collections in general via `packages/SwingSet/` .. that will have to be enough until we get upgrade ready for use.

Although.. I suppose if you had a swingset-based test in which the virtual/durable collections were used to create some objects in vat A, then you carefully drop all references to them from RAM (leaving only exports and/or refs from virtualized data), then you wait a crank or two, then send a message back into vat A that references those objects.. that would exercise the code that rebuilds the objects from saved data.

The tricky part is knowing that the original objects really got GCed, which userspace doesn't get to see. You could add some temporary debug messages to liveslots.js while testing the test, to convince yourself that GC is really happening.

Running the workers under XS (`config.defaultManagerType = 'xs-worker'`) is a good idea, that makes GC far more predictable.

**Comment by Chris-Hibbert** [comment]
> We don't have a way to exercise that until https://github.com/Agoric/agoric-sdk/issues/1848 / https://github.com/Agoric/agoric-sdk/issues/4382 is done and we can upgrade vats. We've got test coverage on durable objects/collections in general via packages/SwingSet/ .. that will have to be enough until we get upgrade ready for use.

I was more interested in virtualization than durability. Can we page objects out and back in at this point?

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 1]
```diff
@@ -43,210 +43,257 @@ export const makeAddPool = (
...
 
-        const tokenKeywordAmountOut = AmountMath.make(
-          secondaryBrand,
-          calcValueToRemove(
-            liqTokenSupply,
-            pool.getSecondaryAmount().value,
-            liquidityValueIn,
-          ),
-        );
+  const makePool = defineKind(
```
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js:104]: is the best documentation on `defineKind` the comment on `defineKindInternal`  in `virtualObjectManager.js`?   The parameter types don't seem to be declared in a way that's visible to VSCode and WebStorm, so I always have to find the docs before I can tell what arguments should be provided in what order.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 2]
```diff
@@ -43,210 +43,257 @@ export const makeAddPool = (
...
+    'pool',
+    (liquidityZcfMint, poolSeat, secondaryBrand) => {
```
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js:106]: I'm trying to understand `defineKind`.  How does it know what parameters to provide to the `init` function? (i.e. If the pool doesn't declare them somewhere, what rules produce the parameters?)

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 3]
```diff
@@ -43,210 +43,257 @@ export const makeAddPool = (
...
+        poolSeat,
+        liquidityBrand,
+        secondaryBrand,
+        updater,
+        notifier,
+      } = state;
+
+      /** @type {XYKPool} */
+      const pool = {
+        getLiquiditySupply: () => state.liqTokenSupply,
```
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js:138]: why is `liqTokenSupply` access through `state`, but `liquidityIsuser`, `poolSeat`, `updater`, etc. (line 126) are extracted and accessed as top-level values?

**Review by Chris-Hibbert — COMMENTED** [review]
This looks fine and straightforward. I asked one question about how `makeKind` works.

Do we have the ability to demonstrate that virtualization can handle paging objects out and back in again? 

I've asked before and it seems like the responses have assumed I was asking about durability and revival of the contract rather than the pool objects.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 1 above)
Comment (turadg) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js:104]: https://github.com/Agoric/agoric-sdk/pull/4603/commits/48dbf0f1fa8b86ee3b6781c4e200fc16cfa854e0#diff-4d09d78ca33d43e2584cfbf5adc00ed1caac51f1edbc10bae3d3c14320e56d30 will help

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 2 above)
Comment (FUDCo) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js:106]: `defineKind` doesn't "know" the parameters to `init`.  The maker function calls `init` with whatever parameters _it_ was given.

**Comment by FUDCo** [comment]
> At [#4706 (comment)](https://github.com/Agoric/agoric-sdk/pull/4706#issuecomment-1055681522) @Chris-Hibbert asks:
> 
> > I'm expecting to see a test demonstrating that objects can be defined, saved, flushed from memory, and restored with behavior identical to the original. Is there a reason that's not appropriate here?
> 
> @FUDCo , should this PR test that? If so, how? Thanks.

I don't think it should have to be the responsibility of users of the virtual object system to write tests to verify that the virtual object system works as advertised.  The virtual objects implementation itself does have tests for that though.

**Comment by Chris-Hibbert** [comment]
> I don't think it should have to be the responsibility of users of the virtual object system to write tests to verify that the virtual object system works as advertised. The virtual objects implementation itself does have tests for that though.

I'm willing to accept that it works as advertised. As a reviewer of a conversion, I'm trying to gain confidence that the right state has been captured. I'd feel a lot more confident if I saw that objects that had been paged out and in behaved the same as before. Without that I have to have confidence that I've identified all the state that is in use, and that I understand how the parts of makeKind interact in great detail.

**Comment by FUDCo** [comment]
> > I don't think it should have to be the responsibility of users of the virtual object system to write tests to verify that the virtual object system works as advertised. The virtual objects implementation itself does have tests for that though.
> 
> I'm willing to accept that it works as advertised. As a reviewer of a conversion, I'm trying to gain confidence that the right state has been captured. I'd feel a lot more confident if I saw that objects that had been paged out and in behaved the same as before. Without that I have to have confidence that I've identified all the state that is in use, and that I understand how the parts of makeKind interact in great detail

Ah, I misunderstood the intent of what you were trying to get at.  That makes a lot of sense and seems like a thing that's genuinely important to verify.

**Comment by erights** [comment]
@FUDCo 
> Ah, I misunderstood the intent of what you were trying to get at. That makes a lot of sense and seems like a thing that's genuinely important to verify.

Is there a way to verify that? Should this PR do so? How?

**Comment by turadg** [comment]
Done in #5187

**Pull request closed without merging**